core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -64,672 +87,1996 @@
-                    if (result.getBlockedReason().isPresent()) {
-                        switch (result.getBlockedReason().get()) {
-                            case WRITER_SCALING:
-                                // no-op
-                                break;
-                            case WAITING_FOR_SOURCE:
-                                schedulerStats.getWaitingForSource().update(1);
-                                break;
-                            case SPLIT_QUEUES_FULL:
-                                schedulerStats.getSplitQueuesFull().update(1);
-                                break;
-                            case MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE:
-                            case NO_ACTIVE_DRIVER_GROUP:
-                                break;
-                            default:
-                                throw new UnsupportedOperationException("Unknown blocked reason: " + result.getBlockedReason().get());
-                        }
-                    }
+        private static StageExecutions create(
+                QueryStateMachine queryStateMachine,
+                Session session,
+                Metadata metadata,
+                RemoteTaskFactory taskFactory,
+                NodeTaskMap nodeTaskMap,
+                ExecutorService executor,
+                SplitSchedulerStats schedulerStats,
+                SubPlan planTree,
+                boolean summarizeTaskInfo)
+        {
+            ImmutableMap.Builder<StageId, SqlStageExecution> allStages = ImmutableMap.builder();
+            ImmutableList.Builder<SqlStageExecution> coordinatorStagesInTopologicalOrder = ImmutableList.builder();
+            ImmutableList.Builder<SqlStageExecution> distributedStagesInTopologicalOrder = ImmutableList.builder();
+            StageId rootStageId = null;
+            ImmutableMap.Builder<StageId, Set<StageId>> children = ImmutableMap.builder();
+            ImmutableMap.Builder<StageId, StageId> parents = ImmutableMap.builder();
+            for (SubPlan planNode : Traverser.forTree(SubPlan::getChildren).breadthFirst(planTree)) {
+                PlanFragment fragment = planNode.getFragment();
+                SqlStageExecution stageExecution = createSqlStageExecution(
+                        getStageId(session.getQueryId(), fragment.getId()),
+                        fragment,
+                        extractTableInfo(session, metadata, fragment),
+                        taskFactory,
+                        session,
+                        summarizeTaskInfo,
+                        nodeTaskMap,
+                        executor,
+                        schedulerStats);
+                StageId stageId = stageExecution.getStageId();
+                allStages.put(stageId, stageExecution);
+                if (fragment.getPartitioning().isCoordinatorOnly()) {
+                    coordinatorStagesInTopologicalOrder.add(stageExecution);
```

Comment:
Yeah, it's a little non intuitive. But the way the plan is structured is top down. The root node has references to it's children. Since we are following the reference direction it is still a topological order (despite the plan leaves come last)